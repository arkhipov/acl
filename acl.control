# ACL extension
comment = 'data type ACL'
default_version = '1.0.3'
module_pathname = '$libdir/acl'
relocatable = true
