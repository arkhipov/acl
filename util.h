/* -------------------------------------------------------------------------
 *
 * util.h
 *
 * Copyright (c) 2015-2023 Vladislav Arkhipov <vlad@arkhipov.ru>
 *
 * -------------------------------------------------------------------------
 */
#ifndef __UTIL_H__
#define __UTIL_H__

#include "postgres.h"
#include "fmgr.h"

#include "utils/array.h"

bool check_access_extract_args(FunctionCallInfo fcinfo, ArrayType **acl,
							   uint32 *mask, ArrayType **who,
							   bool *implicit_allow, bool extract_who,
							   bool has_who_argument);
bool check_access_text_mask_extract_args(FunctionCallInfo fcinfo,
										 ArrayType **acl, text **mask,
										 ArrayType **who, bool *implicit_allow,
										 bool extract_who,
										 bool has_who_argument);

void merge_acls_extract_args(FunctionCallInfo fcinfo, ArrayType **parent,
							 ArrayType **child, bool *container,
							 bool *deny_first);

#endif
