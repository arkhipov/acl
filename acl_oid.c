/* -------------------------------------------------------------------------
 *
 * acl_oid.c
 *
 * Copyright (c) 2015-2023 Vladislav Arkhipov <vlad@arkhipov.ru>
 *
 * -------------------------------------------------------------------------
 */
#include "postgres.h"
#include "fmgr.h"

#include "miscadmin.h"
#if PG_VERSION_NUM >= 90300
#include "access/htup_details.h"
#endif
#include "access/tupmacs.h"
#include "catalog/pg_authid.h"
#include "catalog/pg_type.h"
#include "utils/builtins.h"
#include "utils/syscache.h"

#include "acl.h"
#include "util.h"

PGDLLEXPORT Datum ace_in(PG_FUNCTION_ARGS);
PGDLLEXPORT Datum ace_out(PG_FUNCTION_ARGS);
PGDLLEXPORT Datum acl_check_access_text_current_user(PG_FUNCTION_ARGS);
PGDLLEXPORT Datum acl_check_access_int4_current_user(PG_FUNCTION_ARGS);
PGDLLEXPORT Datum acl_check_access_text_oid(PG_FUNCTION_ARGS);
PGDLLEXPORT Datum acl_check_access_int4_oid(PG_FUNCTION_ARGS);
PGDLLEXPORT Datum acl_check_access_text_name(PG_FUNCTION_ARGS);
PGDLLEXPORT Datum acl_check_access_int4_name(PG_FUNCTION_ARGS);
PGDLLEXPORT Datum acl_merge(PG_FUNCTION_ARGS);

PG_FUNCTION_INFO_V1(ace_in);
PG_FUNCTION_INFO_V1(ace_out);
PG_FUNCTION_INFO_V1(acl_check_access_text_current_user);
PG_FUNCTION_INFO_V1(acl_check_access_int4_current_user);
PG_FUNCTION_INFO_V1(acl_check_access_text_oid);
PG_FUNCTION_INFO_V1(acl_check_access_int4_oid);
PG_FUNCTION_INFO_V1(acl_check_access_text_name);
PG_FUNCTION_INFO_V1(acl_check_access_int4_name);
PG_FUNCTION_INFO_V1(acl_merge);

typedef struct AclEntryOid
{
	AclEntryBase 	base;
	Oid				who;
} AclEntryOid;

#define ACL_TYPE_ALIGNMENT			'i'
#define ACL_TYPE_LENGTH				sizeof(AclEntryOid)

#define PUBLIC_OID					0

#define DatumGetAclEntryP(x)		((AclEntryOid *) DatumGetPointer(x))
#define PG_GETARG_ACL_ENTRY_P(x)	DatumGetAclEntryP(PG_GETARG_DATUM(x))
#define PG_RETURN_ACL_ENTRY_P(x)	PG_RETURN_POINTER(x)

static const char *parse_who(const char *s, void *opaque);
static void format_who(StringInfo out, intptr_t opaque);

static AclEntryBase *extract_acl_entry_base(void *entry);
static bool who_matches(void *entry, intptr_t who);

static Oid get_role_oid(const char *name, bool missing_ok);

Datum
ace_in(PG_FUNCTION_ARGS)
{
	const char	   *s = PG_GETARG_CSTRING(0);
	AclEntryOid	   *entry;

	entry = palloc0(sizeof(AclEntryOid));

	parse_acl_entry(s, &entry->base, entry, parse_who);

	PG_RETURN_ACL_ENTRY_P(entry);
}

Datum
ace_out(PG_FUNCTION_ARGS)
{
	AclEntryOid	   *entry = PG_GETARG_ACL_ENTRY_P(0);
	StringInfo		out;

	out = makeStringInfo();

	format_acl_entry(out, (intptr_t) entry, &entry->base, format_who);

	PG_RETURN_CSTRING(out->data);
}

Datum
acl_check_access_int4_current_user(PG_FUNCTION_ARGS)
{
	ArrayType	   *acl;
	uint32			mask;
	bool			implicit_allow;
	Oid				who;

	if (!check_access_extract_args(fcinfo, &acl, &mask, NULL, &implicit_allow,
								   false, false))
		PG_RETURN_NULL();

	who = GetUserId();

	PG_RETURN_UINT32(check_access(acl, ACL_TYPE_LENGTH, ACL_TYPE_ALIGNMENT,
								  extract_acl_entry_base, mask,
								  (intptr_t) who, who_matches,
								  implicit_allow));
}

Datum
acl_check_access_text_current_user(PG_FUNCTION_ARGS)
{
	ArrayType	   *acl;
	text		   *mask;
	bool			implicit_allow;
	Oid				who;

	if (!check_access_text_mask_extract_args(fcinfo, &acl, &mask, NULL, &implicit_allow,
											 false, false))
		PG_RETURN_NULL();

	who = GetUserId();

	PG_RETURN_TEXT_P(check_access_text_mask(acl, ACL_TYPE_LENGTH,
											ACL_TYPE_ALIGNMENT,
											extract_acl_entry_base, mask,
											(intptr_t) who, who_matches,
											implicit_allow));
}

Datum
acl_check_access_int4_oid(PG_FUNCTION_ARGS)
{
	ArrayType	   *acl;
	uint32			mask;
	Oid				who;
	bool			implicit_allow;

	if (!check_access_extract_args(fcinfo, &acl, &mask, NULL, &implicit_allow,
								   false, true))
		PG_RETURN_NULL();

	if (PG_ARGISNULL(2))
		PG_RETURN_NULL();

	who = PG_GETARG_OID(2);

	PG_RETURN_UINT32(check_access(acl, ACL_TYPE_LENGTH, ACL_TYPE_ALIGNMENT,
								  extract_acl_entry_base, mask,
								  (intptr_t) who, who_matches,
								  implicit_allow));
}

Datum
acl_check_access_text_oid(PG_FUNCTION_ARGS)
{
	ArrayType	   *acl;
	text		   *mask;
	Oid				who;
	bool			implicit_allow;

	if (!check_access_text_mask_extract_args(fcinfo, &acl, &mask, NULL, &implicit_allow,
											 false, true))
		PG_RETURN_NULL();

	if (PG_ARGISNULL(2))
		PG_RETURN_NULL();

	who = PG_GETARG_OID(2);

	PG_RETURN_TEXT_P(check_access_text_mask(acl, ACL_TYPE_LENGTH,
											ACL_TYPE_ALIGNMENT,
											extract_acl_entry_base, mask,
											(intptr_t) who, who_matches,
											implicit_allow));
}

Datum
acl_check_access_int4_name(PG_FUNCTION_ARGS)
{
	ArrayType	   *acl;
	uint32			mask;
	Name			rolename;
	bool			implicit_allow;
	Oid				who;

	if (!check_access_extract_args(fcinfo, &acl, &mask, NULL, &implicit_allow,
								   false, true))
		PG_RETURN_NULL();

	if (PG_ARGISNULL(2))
		PG_RETURN_NULL();

	rolename = PG_GETARG_NAME(2);
	who = get_role_oid(NameStr(*rolename), false);

	PG_RETURN_UINT32(check_access(acl, ACL_TYPE_LENGTH, ACL_TYPE_ALIGNMENT,
								  extract_acl_entry_base, mask,
								  (intptr_t) who, who_matches,
								  implicit_allow));
}

Datum
acl_check_access_text_name(PG_FUNCTION_ARGS)
{
	ArrayType	   *acl;
	text		   *mask;
	Name			rolename;
	bool			implicit_allow;
	Oid				who;

	if (!check_access_text_mask_extract_args(fcinfo, &acl, &mask, NULL, &implicit_allow,
											 false, true))
		PG_RETURN_NULL();

	if (PG_ARGISNULL(2))
		PG_RETURN_NULL();

	rolename = PG_GETARG_NAME(2);
	who = get_role_oid(NameStr(*rolename), false);

	PG_RETURN_TEXT_P(check_access_text_mask(acl, ACL_TYPE_LENGTH,
											ACL_TYPE_ALIGNMENT,
											extract_acl_entry_base, mask,
											(intptr_t) who, who_matches,
											implicit_allow));
}

Datum
acl_merge(PG_FUNCTION_ARGS)
{
	ArrayType	   *parent;
	ArrayType	   *child;
	bool			container;
	bool			deny_first;

	merge_acls_extract_args(fcinfo, &parent, &child, &container, &deny_first);

	PG_RETURN_ARRAYTYPE_P(merge_acls(parent, child,
									 ACL_TYPE_LENGTH, ACL_TYPE_ALIGNMENT,
									 extract_acl_entry_base,
									 container, deny_first));
}

static const char *
parse_who(const char *s, void *opaque)
{
	char			name[NAMEDATALEN];
	int				len = 0;
	Oid				oid;
	AclEntryOid	   *acl_entry = (AclEntryOid *) opaque;

	if (*s == '#')
	{
		for (++s; *s != '\0' && isalnum((unsigned char) *s); ++s)
		{
			name[len++] = *s;

			if (len >= 9)
				ereport(ERROR,
						(errcode(ERRCODE_INVALID_TEXT_REPRESENTATION),
						 errmsg("invalid ACE who")));
		}

		name[len] = '\0';

		oid = DatumGetObjectId(DirectFunctionCall1(oidin,
											  CStringGetDatum(name)));

		acl_entry->base.flags |= ACE_INVALID;
	}
	else
	{
		bool		in_quotes = false;

		for (;
			 *s != '\0' &&
			 (isalnum((unsigned char) *s) ||
			  *s == '_' ||
			  *s == '"' ||
			  in_quotes);
			 ++s)
		{
			if (*s == '"')
			{
				if (*(s + 1) != '"')
				{
					in_quotes = !in_quotes;
					continue;
				}

				++s;
			}

			if (len >= NAMEDATALEN - 1)
				ereport(ERROR,
						(errcode(ERRCODE_NAME_TOO_LONG),
						 errmsg("identifier too long"),
						 errdetail("Identifier must be less than %d characters.",
								   NAMEDATALEN)));

			name[len++] = *s;
		}

		if (len == 0)
		{
			oid = PUBLIC_OID;
		}
		else
		{
			name[len] = '\0';
			oid = get_role_oid(name, true);

			if (!OidIsValid(oid))
				acl_entry->base.flags |= ACE_INVALID;
		}
	}

	acl_entry->who = oid;

	return s;
}

static void
format_who(StringInfo out, intptr_t opaque)
{
	HeapTuple		htup;
	AclEntryOid	   *entry = (AclEntryOid *) opaque;

	if (entry->who == PUBLIC_OID)
		return;

	htup = SearchSysCache1(AUTHOID, ObjectIdGetDatum(entry->who));
	if (!HeapTupleIsValid(htup))
	{
		appendStringInfo(out, "#%d", entry->who);
	}
	else
	{
		char   *name = NameStr(((Form_pg_authid) GETSTRUCT(htup))->rolname);
		char   *s;
		bool	safe = true;

		for (s = name; *s; ++s)
		{
			if (!isalnum((unsigned char) *s) && *s != '_')
			{
				safe = false;
				break;
			}
		}

		if (!safe)
			appendStringInfoChar(out, '"');

		for (s = name; *s; ++s)
		{
			if (*s == '"')
				appendStringInfoChar(out, '"');

			appendStringInfoChar(out, *s);
		}

		if (!safe)
			appendStringInfoChar(out, '"');

		ReleaseSysCache(htup);
	}
}

static AclEntryBase *
extract_acl_entry_base(void *entry)
{
	return &((AclEntryOid *) entry)->base;
}

static bool
who_matches(void *entry, intptr_t who)
{
	Oid			entry_who = ((AclEntryOid *) entry)->who;

	return entry_who == PUBLIC_OID || entry_who == (Oid) who;
}

static
Oid get_role_oid(const char *name, bool missing_ok)
{
	Oid			oid;

	oid = GetSysCacheOid1(AUTHNAME, CStringGetDatum(name));
	if (!missing_ok && !OidIsValid(oid))
		ereport(ERROR,
				(errcode(ERRCODE_UNDEFINED_OBJECT),
				 errmsg("role \"%s\" does not exist", name)));

	return oid;
}
