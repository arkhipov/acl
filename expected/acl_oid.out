-- set up users
create user acl_test1;
create user "acl test2";
create user """acl=""s,/a//acl_test1";
create user acl_temporary_user;
-- ACL format checks
select ''::ace;
ERROR:  missing ACE type
LINE 1: select ''::ace;
               ^
select 'a'::ace;
ERROR:  missing "/" sign
LINE 1: select 'a'::ace;
               ^
select 'q'::ace;
ERROR:  invalid ACE type: must be one of "ad"
LINE 1: select 'q'::ace;
               ^
select 'a*'::ace;
ERROR:  missing "/" sign
LINE 1: select 'a*'::ace;
               ^
select 'a/'::ace;
ERROR:  missing ACE flags
LINE 1: select 'a/'::ace;
               ^
select 'a/hq'::ace;
ERROR:  invalid ACE flag: must be one of "hpcoi0123456789ABCDEFGHIJKLMNOP"
LINE 1: select 'a/hq'::ace;
               ^
select 'a/h'::ace;
ERROR:  missing "/" sign
LINE 1: select 'a/h'::ace;
               ^
select 'a/h/'::ace;
ERROR:  missing ACE who
LINE 1: select 'a/h/'::ace;
               ^
select 'a/h/='::ace;
ERROR:  missing ACE mask
LINE 1: select 'a/h/='::ace;
               ^
select 'a/h/acl_test1='::ace;
ERROR:  missing ACE mask
LINE 1: select 'a/h/acl_test1='::ace;
               ^
select 'a/h/acl_test1=d,'::ace;
ERROR:  invalid ACE mask: must be one of "scdwr0123456789ABCDEFGHIJKLMNOPQ"
LINE 1: select 'a/h/acl_test1=d,'::ace;
               ^
select 'a/h/acl_test1=dw'::ace;
       ace        
------------------
 a/h/acl_test1=dw
(1 row)

select 'a/ihpc/acl_test1=wdddw'::ace;
         ace         
---------------------
 a/hpci/acl_test1=dw
(1 row)

select 'd/ihpc/acl_test1=wdddw'::ace;
         ace         
---------------------
 d/hpci/acl_test1=dw
(1 row)

select 'a/ihpc/"acl test2"=dw0'::ace;
          ace           
------------------------
 a/hpci/"acl test2"=0dw
(1 row)

-- invalid role
select 'a//blah=d'::ace;
  ace   
--------
 a/x/=d
(1 row)

-- deleted role
create table acl_test (ace ace);
insert into acl_test (ace) values ('a//acl_temporary_user=s');
drop user acl_temporary_user;
select ace::text ~ '#' from acl_test;
 ?column? 
----------
 t
(1 row)

-- invalid acl
select (ace::text::ace::text) ~ '#' from acl_test;
 ?column? 
----------
 t
(1 row)

-- current user
set role 'acl_test1';
select coalesce(acl_check_access('{}'::ace[], 'sd0', false), 'NULL');
 coalesce 
----------
 
(1 row)

select acl_check_access('{}'::ace[], (1 << 0) | (1 << 27) | (1 << 29), false)::bit(32);
         acl_check_access         
----------------------------------
 00000000000000000000000000000000
(1 row)

select coalesce(acl_check_access(null::ace[], 'sd0', false), 'NULL');
 coalesce 
----------
 
(1 row)

select acl_check_access(null::ace[], (1 << 0) | (1 << 27) | (1 << 29), false)::bit(32);
         acl_check_access         
----------------------------------
 00000000000000000000000000000000
(1 row)

select acl_check_access('{d//=s,a//acl_test1=sdw}'::ace[], 'sd0', false);
 acl_check_access 
------------------
 d
(1 row)

select acl_check_access('{d//=s,a//acl_test1=sdw}'::ace[], (1 << 0) | (1 << 27) | (1 << 29), false)::bit(32);
         acl_check_access         
----------------------------------
 00100000000000000000000000000000
(1 row)

select acl_check_access('{a//acl_test1=s0,d//=sdw}'::ace[], 'sd0', false);
 acl_check_access 
------------------
 0s
(1 row)

select acl_check_access('{a//acl_test1=s0,d//=sdw}'::ace[], (1 << 0) | (1 << 27) | (1 << 29), false)::bit(32);
         acl_check_access         
----------------------------------
 00001000000000000000000000000001
(1 row)

reset role;
set role """acl=""s,/a//acl_test1";
select acl_check_access('{d/hpc/acl_test1=dw0,a//\"\"\"acl=\"\"s\,/a//acl_test1\"=c}'::ace[], 'c', false);
 acl_check_access 
------------------
 c
(1 row)

select acl_check_access('{d/hpc/acl_test1=dw0,a//\"\"\"acl=\"\"s\,/a//acl_test1\"=c}'::ace[], (1 << 28), false)::bit(32);
         acl_check_access         
----------------------------------
 00010000000000000000000000000000
(1 row)

reset role;
-- name
select coalesce(acl_check_access('{}'::ace[], 'sd0', 'acl_test1', false), 'NULL');
 coalesce 
----------
 
(1 row)

select acl_check_access('{}'::ace[], (1 << 0) | (1 << 27) | (1 << 29), 'acl_test1', false)::bit(32);
         acl_check_access         
----------------------------------
 00000000000000000000000000000000
(1 row)

select coalesce(acl_check_access(null::ace[], 'sd0', 'acl_test1', false), 'NULL');
 coalesce 
----------
 
(1 row)

select acl_check_access(null::ace[], (1 << 0) | (1 << 27) | (1 << 29), 'acl_test1', false)::bit(32);
         acl_check_access         
----------------------------------
 00000000000000000000000000000000
(1 row)

select coalesce(acl_check_access('{}'::ace[], 'sd0', null, false), 'NULL');
 coalesce 
----------
 NULL
(1 row)

select acl_check_access('{}'::ace[], (1 << 0) | (1 << 27) | (1 << 29), null, false)::bit(32);
 acl_check_access 
------------------
 
(1 row)

select acl_check_access('{d//=s,a//acl_test1=sdw}'::ace[], 'sd0', 'acl_test1', false);
 acl_check_access 
------------------
 d
(1 row)

select acl_check_access('{d//=s,a//acl_test1=sdw}'::ace[], (1 << 0) | (1 << 27) | (1 << 29), 'acl_test1', false)::bit(32);
         acl_check_access         
----------------------------------
 00100000000000000000000000000000
(1 row)

select acl_check_access('{a//acl_test1=s0,d//=sdw}'::ace[], 'sd0', 'acl_test1', false);
 acl_check_access 
------------------
 0s
(1 row)

select acl_check_access('{a//acl_test1=s0,d//=sdw}'::ace[], (1 << 0) | (1 << 27) | (1 << 29), 'acl_test1', false)::bit(32);
         acl_check_access         
----------------------------------
 00001000000000000000000000000001
(1 row)

select acl_check_access('{d/hpc/acl_test1=dw0,a//\"\"\"acl=\"\"s\,/a//acl_test1\"=c}'::ace[], 'c', '"acl="s,/a//acl_test1', false);
 acl_check_access 
------------------
 c
(1 row)

select acl_check_access('{d/hpc/acl_test1=dw0,a//\"\"\"acl=\"\"s\,/a//acl_test1\"=c}'::ace[], (1 << 28), '"acl="s,/a//acl_test1', false)::bit(32);
         acl_check_access         
----------------------------------
 00010000000000000000000000000000
(1 row)

-- oid
select coalesce(acl_check_access('{}'::ace[], 'sd0', (select oid from pg_roles where rolname = 'acl_test1'), false), 'NULL');
 coalesce 
----------
 
(1 row)

select acl_check_access('{}'::ace[], (1 << 0) | (1 << 27) | (1 << 29), (select oid from pg_roles where rolname = 'acl_test1'), false)::bit(32);
         acl_check_access         
----------------------------------
 00000000000000000000000000000000
(1 row)

select coalesce(acl_check_access(null::ace[], 'sd0', (select oid from pg_roles where rolname = 'acl_test1'), false), 'NULL');
 coalesce 
----------
 
(1 row)

select acl_check_access(null::ace[], (1 << 0) | (1 << 27) | (1 << 29), (select oid from pg_roles where rolname = 'acl_test1'), false)::bit(32);
         acl_check_access         
----------------------------------
 00000000000000000000000000000000
(1 row)

select coalesce(acl_check_access('{}'::ace[], 'sd0', null, false), 'NULL');
 coalesce 
----------
 NULL
(1 row)

select acl_check_access('{}'::ace[], (1 << 0) | (1 << 27) | (1 << 29), null, false)::bit(32);
 acl_check_access 
------------------
 
(1 row)

select acl_check_access('{d//=s,a//acl_test1=sdw}'::ace[], 'sd0', (select oid from pg_roles where rolname = 'acl_test1'), false);
 acl_check_access 
------------------
 d
(1 row)

select acl_check_access('{d//=s,a//acl_test1=sdw}'::ace[], (1 << 0) | (1 << 27) | (1 << 29), (select oid from pg_roles where rolname = 'acl_test1'), false)::bit(32);
         acl_check_access         
----------------------------------
 00100000000000000000000000000000
(1 row)

select acl_check_access('{a//acl_test1=s0,d//=sdw}'::ace[], 'sd0', (select oid from pg_roles where rolname = 'acl_test1'), false);
 acl_check_access 
------------------
 0s
(1 row)

select acl_check_access('{a//acl_test1=s0,d//=sdw}'::ace[], (1 << 0) | (1 << 27) | (1 << 29), (select oid from pg_roles where rolname = 'acl_test1'), false)::bit(32);
         acl_check_access         
----------------------------------
 00001000000000000000000000000001
(1 row)

select acl_check_access('{d/hpc/acl_test1=dw0,a//\"\"\"acl=\"\"s\,/a//acl_test1\"=c}'::ace[], 'c', (select oid from pg_roles where rolname = '"acl="s,/a//acl_test1'), false);
 acl_check_access 
------------------
 c
(1 row)

select acl_check_access('{d/hpc/acl_test1=dw0,a//\"\"\"acl=\"\"s\,/a//acl_test1\"=c}'::ace[], (1 << 28), (select oid from pg_roles where rolname = '"acl="s,/a//acl_test1'), false)::bit(32);
         acl_check_access         
----------------------------------
 00010000000000000000000000000000
(1 row)

-- inherit only
select acl_check_access('{d/i/=s,a//acl_test1=sdw}'::ace[], 'sd0', 'acl_test1', false);
 acl_check_access 
------------------
 sd
(1 row)

select acl_check_access('{d/i/=s,a//acl_test1=sdw}'::ace[], (1 << 0) | (1 << 27) | (1 << 29), 'acl_test1', false)::bit(32);
         acl_check_access         
----------------------------------
 00101000000000000000000000000000
(1 row)

-- merge
select acl_merge(null::ace[], '{a//=0,d//=1,a//=23,d//=4}'::ace[], true, false);
         acl_merge          
----------------------------
 {a//=0,d//=1,a//=23,d//=4}
(1 row)

select acl_merge(null::ace[], '{a//=0,d//=1,a//=23,d//=4}'::ace[], true, true);
         acl_merge          
----------------------------
 {d//=1,d//=4,a//=0,a//=23}
(1 row)

select acl_merge(null::ace[], '{a//=0,d//=1,a//=23,d//=4}'::ace[], false, false);
         acl_merge          
----------------------------
 {a//=0,d//=1,a//=23,d//=4}
(1 row)

select acl_merge(null::ace[], '{a//=0,d//=1,a//=23,d//=4}'::ace[], false, true);
         acl_merge          
----------------------------
 {d//=1,d//=4,a//=0,a//=23}
(1 row)

-- inheritance
-- container
-- no flags -> not inherited
select acl_merge('{a//acl_test1=d}'::ace[], '{a//=0,d//=1,a//=23,d//=4}'::ace[], true, false);
         acl_merge          
----------------------------
 {a//=0,d//=1,a//=23,d//=4}
(1 row)

-- inherit only -> not inherited
select acl_merge('{a/i/acl_test1=d}'::ace[], '{a//=0,d//=1,a//=23,d//=4}'::ace[], true, false);
         acl_merge          
----------------------------
 {a//=0,d//=1,a//=23,d//=4}
(1 row)

-- object inherit -> inherit only + object inherit
select acl_merge('{a/o/acl_test1=d}'::ace[], '{a//=0,d//=1,a//=23,d//=4}'::ace[], true, false);
                  acl_merge                   
----------------------------------------------
 {a//=0,d//=1,a//=23,d//=4,a/hoi/acl_test1=d}
(1 row)

-- object inherit + inherit only -> inherit only + object inherit
select acl_merge('{a/oi/acl_test1=d}'::ace[], '{a//=0,d//=1,a//=23,d//=4}'::ace[], true, false);
                  acl_merge                   
----------------------------------------------
 {a//=0,d//=1,a//=23,d//=4,a/hoi/acl_test1=d}
(1 row)

-- object inherit + no propagate inherit -> no inheritance
select acl_merge('{a/op/acl_test1=d}'::ace[], '{a//=0,d//=1,a//=23,d//=4}'::ace[], true, false);
         acl_merge          
----------------------------
 {a//=0,d//=1,a//=23,d//=4}
(1 row)

-- object inherit + no propagate inherit + inherit only -> no inheritance
select acl_merge('{a/opi/acl_test1=d}'::ace[], '{a//=0,d//=1,a//=23,d//=4}'::ace[], true, false);
         acl_merge          
----------------------------
 {a//=0,d//=1,a//=23,d//=4}
(1 row)

--container inherit -> container inherit
select acl_merge('{a/c/acl_test1=d}'::ace[], '{a//=0,d//=1,a//=23,d//=4}'::ace[], true, false);
                  acl_merge                  
---------------------------------------------
 {a//=0,d//=1,a//=23,d//=4,a/hc/acl_test1=d}
(1 row)

--container inherit + inherit only -> container inherit
select acl_merge('{a/ci/acl_test1=d}'::ace[], '{a//=0,d//=1,a//=23,d//=4}'::ace[], true, false);
                  acl_merge                  
---------------------------------------------
 {a//=0,d//=1,a//=23,d//=4,a/hc/acl_test1=d}
(1 row)

-- container inherit + no propagate inherit -> no flags
select acl_merge('{a/cp/acl_test1=d}'::ace[], '{a//=0,d//=1,a//=23,d//=4}'::ace[], true, false);
                 acl_merge                  
--------------------------------------------
 {a//=0,d//=1,a//=23,d//=4,a/h/acl_test1=d}
(1 row)

-- container inherit + no propagate inherit + inherit only -> no flags
select acl_merge('{a/cpi/acl_test1=d}'::ace[], '{a//=0,d//=1,a//=23,d//=4}'::ace[], true, false);
                 acl_merge                  
--------------------------------------------
 {a//=0,d//=1,a//=23,d//=4,a/h/acl_test1=d}
(1 row)

-- container inherit + object inherit -> container inherit + object inherit
select acl_merge('{a/co/acl_test1=d}'::ace[], '{a//=0,d//=1,a//=23,d//=4}'::ace[], true, false);
                  acl_merge                   
----------------------------------------------
 {a//=0,d//=1,a//=23,d//=4,a/hco/acl_test1=d}
(1 row)

-- container inherit + object inherit + inherit only -> container inherit + object inherit
select acl_merge('{a/coi/acl_test1=d}'::ace[], '{a//=0,d//=1,a//=23,d//=4}'::ace[], true, false);
                  acl_merge                   
----------------------------------------------
 {a//=0,d//=1,a//=23,d//=4,a/hco/acl_test1=d}
(1 row)

-- container inherit + object inherit + no propagate inherit -> no flags
select acl_merge('{a/cop/acl_test1=d}'::ace[], '{a//=0,d//=1,a//=23,d//=4}'::ace[], true, false);
                 acl_merge                  
--------------------------------------------
 {a//=0,d//=1,a//=23,d//=4,a/h/acl_test1=d}
(1 row)

-- container inherit + object inherit + no propagate inherit + inherit only -> no flags
select acl_merge('{a/copi/acl_test1=d}'::ace[], '{a//=0,d//=1,a//=23,d//=4}'::ace[], true, false);
                 acl_merge                  
--------------------------------------------
 {a//=0,d//=1,a//=23,d//=4,a/h/acl_test1=d}
(1 row)

-- object
-- no flags -> not inherited
select acl_merge('{a//acl_test1=d}'::ace[], '{a//=0,d//=1,a//=23,d//=4}'::ace[], false, false);
         acl_merge          
----------------------------
 {a//=0,d//=1,a//=23,d//=4}
(1 row)

-- inherit only -> not inherited
select acl_merge('{a/i/acl_test1=d}'::ace[], '{a//=0,d//=1,a//=23,d//=4}'::ace[], false, false);
         acl_merge          
----------------------------
 {a//=0,d//=1,a//=23,d//=4}
(1 row)

-- object inherit -> no flags
select acl_merge('{a/o/acl_test1=d}'::ace[], '{a//=0,d//=1,a//=23,d//=4}'::ace[], false, false);
                 acl_merge                  
--------------------------------------------
 {a//=0,d//=1,a//=23,d//=4,a/h/acl_test1=d}
(1 row)

-- object inherit + inherit only -> no flags
select acl_merge('{a/oi/acl_test1=d}'::ace[], '{a//=0,d//=1,a//=23,d//=4}'::ace[], false, false);
                 acl_merge                  
--------------------------------------------
 {a//=0,d//=1,a//=23,d//=4,a/h/acl_test1=d}
(1 row)

-- object inherit + no propagate inherit -> no flags
select acl_merge('{a/op/acl_test1=d}'::ace[], '{a//=0,d//=1,a//=23,d//=4}'::ace[], false, false);
                 acl_merge                  
--------------------------------------------
 {a//=0,d//=1,a//=23,d//=4,a/h/acl_test1=d}
(1 row)

-- object inherit + no propagate inherit + inherit only -> no flags
select acl_merge('{a/opi/acl_test1=d}'::ace[], '{a//=0,d//=1,a//=23,d//=4}'::ace[], false, false);
                 acl_merge                  
--------------------------------------------
 {a//=0,d//=1,a//=23,d//=4,a/h/acl_test1=d}
(1 row)

--container inherit -> not inherited
select acl_merge('{a/c/acl_test1=d}'::ace[], '{a//=0,d//=1,a//=23,d//=4}'::ace[], false, false);
         acl_merge          
----------------------------
 {a//=0,d//=1,a//=23,d//=4}
(1 row)

--container inherit + inherit only -> not inherited
select acl_merge('{a/ci/acl_test1=d}'::ace[], '{a//=0,d//=1,a//=23,d//=4}'::ace[], false, false);
         acl_merge          
----------------------------
 {a//=0,d//=1,a//=23,d//=4}
(1 row)

-- container inherit + no propagate inherit -> not inherited
select acl_merge('{a/cp/acl_test1=d}'::ace[], '{a//=0,d//=1,a//=23,d//=4}'::ace[], false, false);
         acl_merge          
----------------------------
 {a//=0,d//=1,a//=23,d//=4}
(1 row)

-- container inherit + no propagate inherit + inherit only -> not inherited
select acl_merge('{a/cpi/acl_test1=d}'::ace[], '{a//=0,d//=1,a//=23,d//=4}'::ace[], false, false);
         acl_merge          
----------------------------
 {a//=0,d//=1,a//=23,d//=4}
(1 row)

-- container inherit + object inherit -> no flags
select acl_merge('{a/co/acl_test1=d}'::ace[], '{a//=0,d//=1,a//=23,d//=4}'::ace[], false, false);
                 acl_merge                  
--------------------------------------------
 {a//=0,d//=1,a//=23,d//=4,a/h/acl_test1=d}
(1 row)

-- container inherit + object inherit + inherit only -> no flags
select acl_merge('{a/coi/acl_test1=d}'::ace[], '{a//=0,d//=1,a//=23,d//=4}'::ace[], false, false);
                 acl_merge                  
--------------------------------------------
 {a//=0,d//=1,a//=23,d//=4,a/h/acl_test1=d}
(1 row)

-- container inherit + object inherit + no propagate inherit -> no flags
select acl_merge('{a/cop/acl_test1=d}'::ace[], '{a//=0,d//=1,a//=23,d//=4}'::ace[], false, false);
                 acl_merge                  
--------------------------------------------
 {a//=0,d//=1,a//=23,d//=4,a/h/acl_test1=d}
(1 row)

-- container inherit + object inherit + no propagate inherit + inherit only -> no flags
select acl_merge('{a/copi/acl_test1=d}'::ace[], '{a//=0,d//=1,a//=23,d//=4}'::ace[], false, false);
                 acl_merge                  
--------------------------------------------
 {a//=0,d//=1,a//=23,d//=4,a/h/acl_test1=d}
(1 row)

-- skip inherited
select acl_merge('{a/h/acl_test1=d}'::ace[], '{a//=0,d//=1,a//=23,d//=4}'::ace[], false, false);
         acl_merge          
----------------------------
 {a//=0,d//=1,a//=23,d//=4}
(1 row)

-- clean up
drop user acl_test1;
drop user "acl test2";
drop user """acl=""s,/a//acl_test1";
